lib/SILOptimizer/Transforms/DeadObjectElimination.cpp

```diff
@@ -115,13 +135,13 @@ static bool doesDestructorHaveSideEffects(AllocRefInst *ARI) {
   LLVM_DEBUG(llvm::dbgs() << "    Analyzing destructor.\n");
 
   // For each BB in the destructor...
-  for (auto &BB : *Fn)
+  for (auto &BB : *Fn) {
     // For each instruction I in BB...
     for (auto &I : BB) {
       LLVM_DEBUG(llvm::dbgs() << "        Visiting: " << I);
 
       // If I has no side effects, we can ignore it.
-      if (!I.mayHaveSideEffects()) {
+      if (!I.mayHaveSideEffects() || isa<FixLifetimeInst>(&I)) {
```

Comment:
I think we need to justify ignoring fix_lifetime here in a comment. If this were something like a 'dependent_lifetime' instruction, it could be ignored since all references used in the deinitializer will be deleted.

Can we can see that it's a fix_lifetime of 'self'? Then there's no problem deleting it.

If it's a fix_lifetime of something indirectly loaded then I think we should have a TODO explaining that it's probably ok to ignore in practice, but once the deinitializer in question is rewritten to use a more precise dependent_lifetime, then we should go back to treating fix_lifetime conservatively. It's seems possible for someone to use Builtin.fixLifetime in a deinitializer to keep something else alive, just unlikely enough that I'm not really worried.

Reply:
I see what you mean. I changed it to check if the operand is "self".